### ISSUE TITLE
Add mixed precision option to BaseTrainer

### ISSUE THREAD

**Issue by mosheliv** [KevinMusgrave/pytorch-metric-learning#319]
First, thank you for the wonderful library. It helped me achieve amazing results in my project, and your continious support is much appreciated!

I am a it struggling with adding mixed precision training to the example code. Would it be possible to publish a version of the marginloss notebook with mixed precision training?

**Comment by KevinMusgrave** [comment]
I've tried to make all the losses and miners compatible with mixed precision training, so you should be able to follow the [PyTorch example](https://pytorch.org/docs/stable/notes/amp_examples.html#typical-mixed-precision-training). You don't need to do anything special with the pytorch-metric-learning functions. Just call the loss function inside of the ```with autocast():``` block.

**Comment by mosheliv** [comment]
I am using the trainer.train, as per the example. I guess I could copy it
and modify but it would have been nice if it had an FP16 flag

**Comment by KevinMusgrave** [comment]
Ah ok, fair point.

**Comment by mosheliv** [comment]
I am still stuck on this. tried to add with autocast before calling the trainer but this didn't work. looking at the code, its seems that there was no scaler or autocast support in the trainer. I tried adding it but got lost a bit in the maze of calls. if you can give me a hint where the forward is I can try adding it. I should be relatively easy.
Or, if you can post a non trainer training loop, this will work as well for me.
Thank you!!!!

**Comment by KevinMusgrave** [comment]
Maybe this will work:

```python
from .base_trainer import MetricLossOnly
import torch

class WithAutocastMetricLossOnly(MetricLossOnly):
    def calculate_loss(self, curr_batch):
        data, labels = curr_batch
        with torch.cuda.amp.autocast():
            embeddings = self.compute_embeddings(data)
            indices_tuple = self.maybe_mine_embeddings(embeddings, labels)
            self.losses["metric_loss"] = self.maybe_get_metric_loss(
                embeddings, labels, indices_tuple
            )

```

**Comment by mosheliv** [comment]
Yayyyy!!!! it worked! so happy!

Thank you!

**Comment by KevinMusgrave** [comment]
Great!
